### DISCUSSION TITLE
Opening 5250 terminals seems to open PASE terminal instead

### DISCUSSION THREAD

**Discussion by dlanz** [codefori/vscode-ibmi#2372] [General]
Running `which tn5250` returns:
`/QOpenSys/pkgs/bin/tn5250`

Running `tn5250 -version` returns:
`tn5250 0.17.6`

Using the same profile, I'm able to connect with Code for IBM i and open a PASE terminal without issue. However, it seems when attempting to open a 5250 session, a "green screen" flashes momentarily before switching to a PASE terminal. In the resulting PASE terminal, it contains `TERM=xterm /QOpenSys/pkgs/bin/tn5250   env.DEVNAME=PCS2561_4 localhost` in the session history. 

The flash is quick, but long enough to be certain it only occurs when trying to open a 5250 terminal. Opening PASE directly does not produce the behavior.

The system I'm connecting to requires a device description to exist before connecting. I have it configured to prompt for the device name, and I enter a name which is successful when using ACS.

I have tried disconnecting, then using "Connect and Reload Server Settings." as mentioned in the docs, but it doesn't seem to make a difference.

Anyone have any ideas? It would be amazing if I didn't have to switch back and forth between ACS's 5250 and VSCode!

**Comment by worksofliam** [comment]
Interesting. If you run the command yourself from a PASE shell, do you get the same result?

```sh
TERM=xterm /QOpenSys/pkgs/bin/tn5250   env.DEVNAME=PCS2561_4 localhost
```

**Reply by dlanz** [reply]
This is the output I get when I run `TERM=xterm /QOpenSys/pkgs/bin/tn5250   env.DEVNAME=PCS2561_4` from the PASE shell provided through Code for IBM i:

```
tn5250 - TCP/IP 5250 emulator
Syntax:
  tn5250 [options] HOST[:PORT]
   To connect using ssl prefix HOST with 'ssl:'.  Example:      tn5250 +ssl_verify_server ssl:as400.example.com

Options:
   map=NAME                Character map (default is '37'):
                             37, 256, 273, 277, 278, 
                             280, 284, 285, 297, 500, 
                             871, 870, 905, 880, 420, 
                             875, 424, 1026, 290, win37, 
                             win256, win273, win277, win278, win280, 
                             win284, win285, win297, win500, win871, 
                             win870, win905, win880, win420, win875, 
                             win424, win1026, 
   env.DEVNAME=NAME         Use NAME as session name (default: none).
   trace=FILE              Log session to FILE.
   +/-ssl_verify_server    Verify/don't verify the server's SSL certificate
   ssl_ca_file=FILE        Use certificate authority (CA) certs from FILE
   ssl_cert_file=FILE      File containing SSL certificate in PEM format to
                           use if the AS/400 requires client authentication.
   ssl_pem_pass=PHRASE     Passphrase to use when decrypting a PEM private
                           key.  Used in conjunction with ssl_cert_file
   ssl_check_exp[=SECS]    Check if SSL certificate is expired, or if it
                           will be expired in SECS seconds.
   +/-underscores          Use/don't use underscores instead of underline
                           attribute.
   +/-ruler                Draw a ruler pointing to the cursor position
   +/-version              Show emulator version and exit.
   env.NAME=VALUE          Set telnet environment string NAME to VALUE.
   env.TERM=TYPE           Emulate IBM terminal type (default: depends)
                             IBM-3477-FC (27x132 color)
                             IBM-3477-FG (27x132 monochrome)
                             IBM-3180-2 (27x132 monochrome)
                             IBM-3179-2 (24x80 color)
                             IBM-3196-A1 (24x80 monochrome)
                             IBM-5292-2 (24x80 color)
                             IBM-5291-1 (24x80 monochrome)
                             IBM-5251-11 (24x80 monochrome)
```

Does TN5250 require any type of configuration on the system before it can be used? It has only been installed via YUM, nothing further has been done with it on the system.

**Reply by dlanz** [reply]
~~Sorry, I just realized you added `localhost` to the end of that.~~
Sorry, I just realized I omitted the `localhost` in the prior post.

Yes, I get the same behavior of the flash when I run:
`TERM=xterm /QOpenSys/pkgs/bin/tn5250   env.DEVNAME=PCS2561_4 localhost`

**Reply by worksofliam** [reply]
@dlanz Looking at the tn5250 man page, I see you're already passing in `DEVNAME` - I assume device name and description (as you wrote in the original post) are the same?

https://linux.die.net/man/5/tn5250rc

**Reply by dlanz** [reply]
I'm pretty sure. This is what I get after running:
```
WRKDEVD DEVD(PCS2561_4)
```

![Screenshot 2024-11-26 163200](https://github.com/user-attachments/assets/61c735e3-4d11-42c6-8733-c44d3c47a674)

I tried running `TERM=xterm /QOpenSys/pkgs/bin/tn5250   env.DEVNAME=PCS2561_4 localhost` in an SSH Tunnel outside of VSCode and get the same exact behavior. 

So I doubt this is a Code for IBM i specific issue, but I'm a step closer now. I'll try to read up on TN5250 a bit more to try and figure out what's going on. Thank you for the help for getting this far though.

**Reply by worksofliam** [reply]
@chrjorgensen is this something you might know about?

**Reply by chrjorgensen** [reply]
@dlanz @worksofliam I do not get the same behavior on my systems... 😞 

You could try tracing the `tn5250` by adding the option `trace=<tracefile>` - it may give you a hint about what's causing the error in `tn5250`.

**Comment by dlanz** [comment]
Figured I would just update:

I'm not familiar with tn5250 traces, but this is what I'm getting in the trace file and from what I can tell I'm not seeing any errors.
```
Non-Blocking
tn5250_dbuffer_clear_table() entered.
tn5250_char_map_new: map = "37"
Found string for cap '@8': '^[OM'.
Found string for cap 'cr': '^M'.
Found string for cap 'kB': '^[[Z'.
Found string for cap 'k1': '^[OP'.
Found string for cap 'k2': '^[OQ'.
Found string for cap 'k3': '^[OR'.
Found string for cap 'k4': '^[OS'.
Found string for cap 'k5': '^[[15~'.
Found string for cap 'k6': '^[[17~'.
Found string for cap 'k7': '^[[18~'.
Found string for cap 'k8': '^[[19~'.
Found string for cap 'k9': '^[[20~'.
Found string for cap 'k;': '^[[21~'.
Found string for cap 'F1': '^[[23~'.
Found string for cap 'F2': '^[[24~'.
Found string for cap 'F3': '^[[1;2P'.
Found string for cap 'F4': '^[[1;2Q'.
Found string for cap 'F5': '^[[1;2R'.
Found string for cap 'F6': '^[[1;2S'.
Found string for cap 'F7': '^[[15;2~'.
Found string for cap 'F8': '^[[17;2~'.
Found string for cap 'F9': '^[[18;2~'.
Found string for cap 'FA': '^[[19;2~'.
Found string for cap 'FB': '^[[20;2~'.
Found string for cap 'FC': '^[[21;2~'.
Found string for cap 'FD': '^[[23;2~'.
Found string for cap 'FE': '^[[24;2~'.
Found string for cap 'kl': '^[OD'.
Found string for cap 'kr': '^[OC'.
Found string for cap 'ku': '^[OA'.
Found string for cap 'kd': '^[OB'.
Found string for cap 'kP': '^[[5~'.
Found string for cap 'kN': '^[[6~'.
Found string for cap 'kb': '^H'.
Found string for cap 'kh': '^[OH'.
Found string for cap '@7': '^[OF'.
Found string for cap 'kI': '^[[2~'.
Found string for cap 'kD': '^[[3~'.
Macro: fname=/home/DL/.tn5250macros
HOST, This->status  = 8 0
GotVerb(2):<IAC><DO><NEWENV>
GotVerb(3):<IAC><DO><NEWENV>
HOST, This->status  = 8 0
GotVerb(2):<IAC><DO><TERMTYPE>
GotVerb(3):<IAC><DO><TERMTYPE>
GotSB:<IAC><SB><NEWENV><SEND>
        <USERVAR>IBMRSEED<E3 D5 04 94 1B C8 F8 15>
        <VAR><USERVAR><IAC><SE>
SentSB:<IAC><SB><NEWENV><IS>
        <VAR>DEVNAME
        <USERVAR>PCS2561_4
        <VAR>TERM
        <USERVAR>IBM-3179-2<IAC><SE>
GotSB:<IAC><SB><TERMTYPE><SEND><IAC><SE>
SentSB:<IAC><SB><TERMTYPE><IS>IBM-3179-2<IAC><SE>
HandleReceive: entered.
HOST, This->status  = 8 1
GotVerb(2):<IAC><DO><END_OF_REC>
GotVerb(3):<IAC><DO><END_OF_REC>
HOST, This->status  = 8 1
GotVerb(2):<IAC><WILL><END_OF_REC>
GotVerb(3):<IAC><WILL><END_OF_REC>
HOST, This->status  = 8 1
GotVerb(2):<IAC><DO><BINARY>
GotVerb(3):<IAC><DO><BINARY>
HOST, This->status  = 8 1
GotVerb(2):<IAC><WILL><BINARY>
GotVerb(3):<IAC><WILL><BINARY>
HandleReceive: entered.
GotSB:<IAC><SB><NEWENV><SEND>
        <USERVAR>DEVNAME<IAC><SE>
SentSB:<IAC><SB><NEWENV><IS>
        <VAR>DEVNAME
        <USERVAR>PCS2561_4
        <VAR>TERM
        <USERVAR>IBM-3179-2<IAC><SE>
```

The "flashing" behavior, I realized it almost looks as if it is successfully connecting, but then immediately disconnecting. I feel like I can briefly make out what I believe to be a 24x80 screen within the session. I'm still trying to read up on configuring tn5250 on the server. I strongly suspect it's a setting on our server, just not sure which one. If I figure it out, I'll update here.

**Comment by dlanz** [comment]
I stumbled on the issue. 

So I knew the device description could only be used by a single session, so I had logged out of the PCS2561_4 session via ACS before attempting any connections with TN5250. However, I only logged out of the system and left the window opened at the sign on screen. 

I have no idea why I didn't consider sitting at the sign on screen as an open connection. As far as telnet and the system is concerned, it is still an open connection under the device description.

**Reply by chrjorgensen** [reply]
@dlanz Yes, a device showing the login screen is an active device and cannot be acquired by another session. You get the same result, if you try to connect to that specific, active device in another ACS 5250 session. So not a TN5250 issue - but it could be nice and report the cause instead of just dropping out! 🥲 Maybe an idea for the TN5250 project?

**Reply by dlanz** [reply]
@chrjorgensen Yep, certainly just a picnic error (Problem in chair, not in computer 😄).

That's a good idea, I'll take a look on my lunch break and see if I can make a PR on it.
